### DISCUSSION TITLE
Beginner Help: Printing Midi Messages on SH1106 Display

### DISCUSSION THREAD

**Discussion by tilllt** [tttapa/Control-Surface#986] [General]
Hey there,

quick intro:
my longterm dream would be to connect my Polyend Tracker and my Boss BR-800 (and possibly other Devices like my Pocket Operators) wirelessly, so i can run them in sync, without "computers" involved. 

But i am taking it slowly: I started off with a Arduino Nano 33 BLE, but i learned that BLE is not supported with the Nano 33 (yet) by Control Surface, so now i have set up a Lilygo ESP32-S2 with a generic Aliexpress SH1106 SPI Display, i cant get the U8G2 library to work but both the Pololu SH1106 and the Adafruit SH1106 Libraries work.

So as a first step, i would like to use the ESP32 in USB Host Mode, connect the BR-800 and display its Midi Messages (MTC would be lovely) on the SH1106 display - since the USB port is occupied by the BR-800 i cannot use serial monitor. I have some ESP32-C3 with dual USB Ports on the way, but i dont know when they will arrive and if they will cause other problems that i cannot foresee now.

I started browsing the docs the examples, but when it comes to adding a currently unsupported Display (as far as i can see only SSD1306 is supported right now), and sending the "mididbg" output to anything else but the serial port, i am lost.

Would appreciate any pointers to advance with this!

**Comment by tttapa** [comment]
> i would like to use the ESP32 in USB Host Mode

Unfortunately, this is not currently supported. The ESP32-S2 uses TinyUSB for its USB stack, and TinyUSB only supports MIDI in Device mode, not Host mode. Implementing MIDI Host mode requires implementing a custom device driver, using the low-level TinyUSB and ESP-IDF APIs, and is not a trivial task, because you need to have good knowledge of the entire USB spec. Device mode is much simpler.  
To further complicate matters, it seems that the Boss BR-800 can act as both a MIDI device and an audio interface. If you're lucky, those are two Interfaces on a single USB device, if you're unlucky, they're two separate devices behind a USB Hub. Your USB drivers need to be able to deal with such a topology.

See [MIDI over USB](https://tttapa.github.io/Control-Surface-doc/Doxygen/d8/d4a/md_pages_MIDI-over-USB.html) for a table of supported boards.

The only Arduino-compatible boards with good USB MIDI Host support that I'm aware of are the Teensy 3.6 and 4.1.

> display its Midi Messages (MTC would be lovely) on the SH1106 display

That's certainly doable. See [MIDI-Monitor-OLED.ino](https://tttapa.github.io/Control-Surface-doc/Doxygen/d2/d77/MIDI-Monitor-OLED_8ino-example.html). Switching from Adafruit_SSD1306 to Adafruit_SH1106 should be a trivial change, it's unrelated to the Control Surface library.

> since the USB port is occupied by the BR-800 i cannot use serial monitor.

In that case, I'd highly recommend getting one of those cheap USB-to-UART converters (e.g. one based on the FTDI-232RL chip). Connect it to e.g. `Serial1` and use that for the serial monitor.  
Alternatively, get an ESP32 dev board with such a chip built-in (e.g. the ESP32-S3 devkit by Espressif).

> I have some ESP32-C3 with dual USB Ports on the way, but i dont know when they will arrive and if they will cause other problems that i cannot foresee now.

Are you sure you've ordered the ESP32-C3? It does not have USB support (only for Serial and JTAG, but not anything else like MIDI).

> I started browsing the docs the examples, but when it comes to adding a currently unsupported Display (as far as i can see only SSD1306 is supported right now), and sending the "mididbg" output to anything else but the serial port, i am lost.

There is no need to add any unsupported displays. The display operates independently from the Control Surface library. You just use the Adafruit library, directly in your sketch.

**Reply by tilllt** [reply]
Thanks for taking the time, I appreciate it. Seems really difficult to find the supported board for this specific purpose. I like the teensy boards but they are quite expensive in comparison to an average ESP32 dev board...

These are the dual USB ESP32 boards I mentioned:

https://de.aliexpress.com/item/1005004617322170.html

So, to clarify, since I saw this page has a USB Midi Host example for the esp32-s2. But it would not be possible to combine these and make them work with the control surface library? But generally speaking the hardware is capable of acting as a USB midi Host?

https://github.com/touchgadget/esp32-usb-host-demos

**Reply by tttapa** [reply]
Sure, if the code in the esp32-usb-host-demos repository works for your BR-800, it should be trivial to integrate with Control Surface. You can use the [Custom-USB-MIDI-Backend.ino](https://tttapa.github.io/Control-Surface-doc/Doxygen/d7/d5f/Custom-USB-MIDI-Backend_8ino-example.html) example to get started.

**Reply by tilllt** [reply]
I found an even more promising project for the ESP32-S3 (probably works with the S2) with a complete USB Midi Host to BLE Sketch in the examples folder, using freeRTOS threading and all:

https://github.com/aselectroworks/AtomS3_USB-BLE_MIDI_Bridge
